feat: Machine "ping" (RTT from corrosion) times

jabr opened the pull request:
Just experimenting with an idea, but I was thinking the round-trip time (RTT) data from corrosion might be useful for some future Uncloud functionality (e.g. an improved "nearest" local DNS that can meaningfully order past machine-local, possibly some integration with Caddy upstream routing, replica placement preference ["spread out", "keep close"], etc).

I also added a `uc machine rtt` command to show the ~avg~median and stddev of RTT times for each machine/peer pair, which might be a useful diagnostic tool all on its own. Output is something like this:
```
MACHINE    PEER       MEDIAN   STDDEV
amalthea   himalia    140ms    ±19.4ms
amalthea   thebe      39ms     ±1.1ms
himalia    amalthea   168ms    ±18.5ms
himalia    thebe      203ms    ±42.3ms
thebe      amalthea   40ms     ±2.0ms
thebe      himalia    158ms    ±15.2ms```

Also added `RTT` column to `wg show` output.

jabr committed:
Add corrosion admin client function to get RTT to other machines in the cluster

jabr committed:
Add `uc machine rtt` showing all pair-wise RTT stats from corrosion

jabr committed:
Add long description to machine rtt command

jabr commented:
> did not add it to InspectMachine was because that function is heavily used other commands

Looking at the MR again, I realized I had mixed up `ListMachines` and `InspectMachine` and thought you were suggesting adding it to `ListMachines`...

Looking closer at `InspectMachine`, that makes a lot more sense. I'll add it to that.

jabr committed:
Include machine peer RTTs in InspectMachine instead of adding new gRPC API for it

jabr committed:
Add nil check on InspectMachine's Rtt field to protect from a potential edge case

jabr committed:
Generate docs for machine rtt command

jabr committed:
Merge branch 'main' into machine-rtts

jabr committed:
Handle m.Message being nil when a node is down or unavailble

jabr committed:
Update cli-docs

psviderski commented on internal/corrosion/admin.go:
```diff
@@ -256,3 +257,97 @@ func (c *AdminClient) ClusterMembershipStates(latest bool) ([]ClusterMembershipS
+
+type MemberRTTStats struct {
+	Addr    netip.AddrPort
+	Average float64
```
What are the units for rtts? Are they milliseconds?

jabr replied:
Yeah, milliseconds. https://github.com/superfly/corrosion/blob/1b0fcf1b12a81c5394dd4b9e5a7815b0ab95d575/crates/corro-types/src/members.rs#L176

psviderski commented on cmd/uncloud/machine/rtt.go:
```diff
@@ -0,0 +1,104 @@
+	fmt.Fprintln(tw, "MACHINE\tPEER\tRTT")
+
+	for _, r := range rows {
+		fmt.Fprintf(tw, "%s\t%s\t%.1f ±%.1f\n", r.machine, r.peer, r.avg, r.stdDev)
```
What do you think about adding units to the numbers so it's clear what they mean?

psviderski commented on internal/corrosion/admin.go:
```diff
@@ -256,3 +257,97 @@ func (c *AdminClient) ClusterMembershipStates(latest bool) ([]ClusterMembershipS
+		for _, rtt := range rtts {
+			sum += rtt
+		}
+		avg := sum / float64(len(rtts))
```
What are your considerations regarding using avg vs median?
I guess there could be a network hiccup and one data point could be significantly larger than others which may skew the average. Is this what we typically want or the median would better suit us e.g. for sorting upstreams based on the latency (rtt)?

jabr replied:
Makes sense. That’s basically the same reason I included the standard deviation, but median would be a better single number metric to have for most applications using this (eg routing, “nearest” dns, etc).

Should I keep mean/stddev, too? Or just replace both of those with median?

psviderski replied:
Up to you. I think the standard deviation is still useful for detecting unstable links (high jitter). Not sure if we end up using it for nearest detection but might be useful for network troubleshooting. But IMO the mean itself doesn't add value once you have the median.

jabr replied:
I went with median and stddev. New output table looks like:
```
MACHINE    PEER       MEDIAN   STDDEV
amalthea   himalia    140ms    ±19.4ms
amalthea   thebe      39ms     ±1.1ms
himalia    amalthea   168ms    ±18.5ms
himalia    thebe      203ms    ±42.3ms
thebe      amalthea   40ms     ±2.0ms
thebe      himalia    158ms    ±15.2ms
```

psviderski commented on cmd/uncloud/machine/rtt.go:
```diff
@@ -0,0 +1,104 @@
+	})
+
+	// Print table
+	tw := tabwriter.NewWriter(os.Stdout, 0, 0, 3, ' ', 0)
```
Please use https://github.com/psviderski/uncloud/blob/main/internal/cli/tui/table.go#L9-L24. See other commands have been consistently migrated to it.

jabr replied:
Yeah, I saw those updates last night when rebasing the proxy branch. I should have time tonight to update this and make the millisecond label and median vs avg/stddev changes.

psviderski reviewed (approved):
This looks really good! Sorry it took that long for me to take a look at this PR.
I actually need this for my multi-region setup to sort upstreams in Caddy and route traffic to the local region first on each Caddy instance.

Please see a few questions and minor comments below.

jabr committed:
Merge branch 'main' into machine-rtts

jabr committed:
Use tui Table instead of tabwriter

psviderski commented:
Just another thought after sleeping on it. Did you consider displaying RTTs in the output of any existing commands?
I'm thinking that even though we take this data from the Corrosion cluster members (machines), it's about the network layer. And the network layer in uncloud is WireGuard.

We have the `uc wg show` command that lists all peers of the machine (currently connected to or `--machine`). It feels very natural to add the RTT column along HANDSHAKE, RECEIVED, SENT there to see all the relevant data in one place.

The only difference with the `machine rtt` command is that it prints all the combinations of machine <-> peer.

I'm still happy to merge this one as is, just food for thought to potentially improve the UX.

jabr committed:
Change reported RTT to the median rather than mean. Still calculate include stddev, as it might be a useful indication of network jitter.

jabr committed:
Show RTT to peers in `wg show` output

jabr commented:
> We have the uc wg show command ... It feels very natural to add the RTT column ...

Good idea! I added a RTT column to the `wg show` output, too. Very natural fit for that command.

```
PEER       PUBLIC KEY   ENDPOINT          HANDSHAKE   RTT     RECEIVED   SENT      ALLOWED IPS
amalthea   9dK...Oio=   65.x.x.x:51820    22s ago     121ms   735.3MB    764.7MB   fdcc:xxx/128, 10.210.1.0/24
thebe      1Wg...yFI=   93.x.x.x:51820   1m14s ago   138ms   839.1MB    774.4MB   fdcc:xxxx/128, 10.210.0.0/24
```

I also thought about other commands that might make sense, but none felt like a great fit, imo. `machine ls` is the closest, but it doesn't have the "from a particular machine" perspective needed for the directional RTT to make sense.

`wg show` has the "from connection machine" perspective, and `machine rtt` shows all pair-wise combos of machines, so it's an "omni" perspective. `machine ls` is technically a view from the connection machine, as well, but that's more of an implementation detail than something obvious/emphasized to the user like `wg show` is.

jabr committed:
Update cli docs

jabr commented:
Everything from the review should be updated now and ready for another look. Added `ms` unit to output (and improved table column labels), using median instead of average, moved to `tui.Table`, and added RTT to `wg show`.

Let me know if you see anything else.
